internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -73,8 +73,49 @@ object ClassToAPI {
                  new mutable.HashSet,
                  new mutable.HashSet)
 
-  def classCanonicalName(c: Class[_]): String =
-    Option(c.getCanonicalName) getOrElse { c.getName }
+  /**
+   * The binary name encoding for Scala has been broken for a while. It does not
+   * follow the Java source name rules because it does not append a `$` after the
+   * mirror class of an object, which already ends in `$`.
+   *
+   * Zinc works around this fix in the only way it can: it gets the enclosing class,
+   * which is obtained from the `InnerClasses` java classfile metadata, and it
+   * reconstructs the path from there. It circumvents, by design, the core issue:
+   * it doesn't call Java's name parser at all.
+   *
+   * This issue has been fixed in the JDK9, but unfortunately it's going to be tormenting
+   * us for a while because 2.13 does not target Java 9, and we're still at 2.12.x.
+   *
+   * <a href="http://bugs.java.com/bugdatabase/view_bug.do?bug_id=8057919">This is the fix in Java 9.</a>
+   *
+   * In order to fully fix this issue, the Scala compiler should do the desired behaviour
+   * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
+   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   *
+   * @return The canonical name if not null, the full name otherwise.
+   */
+  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+    if (c == null) "" // Return nothing if it hits the top-level class
```

Comment:
Proof at the bytecode level (see ifnotnull):
```
    Code:
      stack=2, locals=1, args_size=1
         0: aload_0
         1: ldc           #71                 // String
         3: putfield      #66                 // Field a:Ljava/lang/String;
         6: aload_0
         7: invokevirtual #73                 // Method a:()Ljava/lang/String;
        10: ifnonnull     24
        13: getstatic     #78                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
        16: ldc           #80                 // String HEY
        18: invokevirtual #84                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
        21: goto          32
        24: getstatic     #78                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
        27: ldc           #86                 // String HOY
        29: invokevirtual #84                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
        32: return
```
  
Compiling with `eq null` produces the same bytecode (`ifnonnull`).